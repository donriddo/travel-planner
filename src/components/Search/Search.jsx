import React from 'react';
import { connect } from 'react-redux';
import qs from 'query-string';
import { geocodeByAddress } from 'react-places-autocomplete';

import { searchActions } from '../../_actions';

import Autocomplete from '../general/Autocomplete';

class Search extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            date: null, passengers: 0,
        };
      }
     
      handleChange(field, value) {
        this.props.dispatch(searchActions.updateField({ [field]: value }));
      };

      handleGenericChange(field, evt) {
        this.setState({ [field]: evt.target.value });
      }
     
      handleStartSelect(address) {
        this.props.dispatch(searchActions.geocodeStartAddress(address));
      };

      handleEndSelect(address) {
        this.props.dispatch(searchActions.geocodeEndAddress(address));
      };

      handleSubmit(evt) {
        evt.preventDefault();
        const data = this.props.search;
        data.passengers = this.state.passengers;
        data.date = this.state.date;
        this.props.dispatch(searchActions.getDistance(data));
      }

    componentDidMount() {
        const { location: { search } } = this.props;
        const queries = qs.parse(search);
        if (queries.start) this.handleStartSelect(queries.start);
        if (queries.end) this.handleEndSelect(queries.end);
    }

    render() {
        const searchOptions = { types: ['geocode'] };
        const { start, end } = this.props.search;
        return (
            <div>
                <form>
                    <ul>
                        <li>
                            <Autocomplete
                                value={start || ''}
                                onChange={this.handleChange.bind(this, 'start')}
                                onSelect={this.handleStartSelect.bind(this)}
                                searchOptions={searchOptions}
                                placeholder="Enter your address"
                            />
                        </li>
                        <li>
                            <Autocomplete
                                value={end || ''}
                                onChange={this.handleChange.bind(this, 'end')}
                                onSelect={this.handleEndSelect.bind(this)}
                                searchOptions={searchOptions}
                                placeholder="Enter destination address"
                            />
                        </li>
                        <li><input onChange={this.handleGenericChange.bind(this, 'passengers')} type="number" placeholder="Number of passengers"/></li>
                        <li><input onChange={this.handleGenericChange.bind(this, 'date')} type="date"/></li>
                        <li><button onClick={this.handleSubmit.bind(this)} type="submit">Get Distance</button></li>
                    </ul>
                    
                </form>
            </div>
        );
    }
}

function mapStateToProps(state) {
    const { search } = state;
    return {
        search
    };
}

export default connect(mapStateToProps)(Search);