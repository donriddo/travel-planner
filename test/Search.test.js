import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

import Search from '../src/components/Search/Search';
import { setupGoogleMock } from './setup';

const mockStore = configureMockStore();
const store = mockStore({
    search: { start: '', end: '', passengers: '', date: '' }
});

beforeAll(() => {
  setupGoogleMock();
});

test('Search test', () => {
  const component = renderer.create(
    <Provider store={store}>
        <Search location={{}} />
    </Provider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});