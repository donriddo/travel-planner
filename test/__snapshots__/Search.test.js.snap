// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search test 1`] = `
<div>
  <form>
    <ul>
      <li>
        <div>
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            autoComplete="off"
            className="location-search-input"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Enter your address"
            role="combobox"
            type="text"
            value=""
          />
          <div
            className="autocomplete-dropdown-container"
          />
        </div>
      </li>
      <li>
        <div>
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            autoComplete="off"
            className="location-search-input"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Enter destination address"
            role="combobox"
            type="text"
            value=""
          />
          <div
            className="autocomplete-dropdown-container"
          />
        </div>
      </li>
      <li>
        <input
          onChange={[Function]}
          placeholder="Number of passengers"
          type="number"
          value=""
        />
      </li>
      <li>
        <input
          onChange={[Function]}
          type="date"
          value=""
        />
      </li>
      <li>
        <button
          onClick={[Function]}
          type="submit"
        >
          Get Distance
        </button>
      </li>
    </ul>
  </form>
</div>
`;
